Enable nss_wrapper in images to enable use of SecurityContext

thallgren opened the pull request:
This commit replaces the `ko` based build with a three-step docker
build. All steps are based on Alpine.

1. `tel2-base` Builds the `nss_wrapper` from source and copies the password
   template and script to activate it.
2. `tel2-build` Builds the `traffic` image from source. The step uses
   `go mod download` to ensure that dependencies are cached by docker
   and then copies the sources needed to build the binary.
3. `tel2` starts with clean alpine image and copies the necessary built
   artifacts from `tel2-base` and `tel2-build`

Everything related to `ko` is removed from the repository.

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->

 - [ ] I made sure to update `./CHANGELOG.md`.
 - [ ] I made sure to add any docs changes required for my change (including release notes).
 - [ ] My change is adequately tested.
 - [ ] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.
 - [ ] I updated `TELEMETRY.md` if I added, changed, or removed a metric name.
 - [ ] Once my PR is ready to have integration tests ran, I posted the PR in #telepresence-dev in the datawire-oss slack so that the "ok to test" label can be applied.

Commit:
Enable nss_wrapper in images to enable use of SecurityContext

This commit replaces the `ko` based build with a three-step docker
build. All steps are based on Alpine.

1. `tel2-base` Builds the `nss_wrapper` from source and copies the password
   template and script to activate it.
2. `tel2-build` Builds the `traffic` image from source. The step uses
   `go mod download` to ensure that dependencies are cached by docker
   and then copies the sources needed to build the binary.
3. `tel2` starts with clean alpine image and copies the necessary built
   artifacts from `tel2-base` and `tel2-build`

Everything related to `ko` is removed from the repository.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Check in the telepresence-chart.tgz in the git repo

This file is embedded and needs to be included in the vendor directory
in downstream components. It must be included in the repository.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

maxlegault commented on base-image/startscript.sh:
```diff
@@ -0,0 +1,11 @@
+#! /bin/sh
```
I think there's an extra space here between the shebang and /bin/sh

maxlegault replied:
And to be safer, we should probably add a `set -eu` to avoid unexpected behaviours for missing variables or continued execution over a previous failure.

maxlegault reviewed (commented):
Minor comment in my review, it looks good to me in general, but I'd like a review from someone who's more used to this code

josecv reviewed (approved):
This looks good to me but have you tested it with istio? I'd just like to be sure that the firewall rules with os.Getuid() won't conflict (they shouldn't bc istio uses a custom UID)

Commit:
In the startup script, remove space after shebang and use set -eu

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Update CHANGELOG.md

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Ensure that telepresence-chart.tgz isn't git-modified on each build

Changes `make build` so that it generates a telepresence-chart.tgz that
is used for that build only (and hence embedded) and then git restored.
A new `make build-version` target is added that doesn't restore the
file. This target is used when pushing a new executable.

The `telepresence-chart.tgz` is also regenerated and committed by the
`prepare-release` target.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Don't run Test_APIServerIsProxied on host with covering default route

When the cluster runs in a VM on the same host as Telepresence and that
VM uses a bridged network, and that bridged network uses the same
default gateway, then it's not possible to "also-proxy" that gateway.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>
